CSR condition status, lastTransitionTime, versioned validation

liggitt opened the pull request:
**What type of PR is this?**
/kind api-change
/kind feature

**What this PR does / why we need it**:

Implements v1beta1 portions of the [CSR KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/20190607-certificates-api.md) (https://github.com/kubernetes/enhancements/issues/1513)

Best reviewed commit-by-commit

* Adds a well-known condition type of `Failed` for use by signers that want to indicate permanent failure
  * Updates the signer controller to record a Failed condition when a request does not pass signerName validation
  * Updates printers/describers to surface failed status
  * Updates cert manager to stop waiting for failed CSRs
  * Updates cert cleaner to clean up stale failed CSRs
  * Allows conditions other than `Approved` and `Denied` to be added via `/status`
* Brings CSR conditions in line with recommended condition schema:
  * Adds a status field (for compatibility, defaults to `True` if unspecified, may only be `True` for `Approved`, `Denied`, and `Failed` conditions)
  * Adds a lastTransitionTime field
* Plumbs API version information to CSR validation
* Partitions validation functions for the main resource, status, and approval subresources
* In all API versions:
  * `Approved` and `Denied` conditions are mutually exclusive (fixes https://github.com/kubernetes/kubernetes/issues/90510)
  * `Approved`, `Denied`, and `Failed` conditions cannot be removed once set
* In versions other than v1beta1:
  * `status.certificate` is required to be well-formed
  * `status.certificate` is immutable once set 
  * duplicate condition types are not allowed
  * empty condition type is not allowed
  * legacy-unknown signer is not allowed

**Does this PR introduce a user-facing change?**:
```release-note
CertificateSigningRequest API conditions were updated:
* a `status` field was added; this field defaults to `True`, and may only be set to `True` for `Approved`, `Denied`, and `Failed` conditions
* a `lastTransitionTime` field was added
* a `Failed` condition type was added to allow signers to indicate permanent failure; this condition can be added via the `certificatesigningrequests/status` subresource.
* `Approved` and `Denied` conditions are mutually exclusive
* `Approved`, `Denied`, and `Failed` conditions can no longer be removed from a CSR
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
```docs
https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/20190607-certificates-api.md
```

/sig auth
/milestone v1.19
/priority important-soon
/cc @deads2k @munnerz

liggitt committed:
Add signerName constants

liggitt committed:
Add conditions status field

liggitt committed:
Plumb version info to validation, separate main/status/approval validation

liggitt committed:
Generated files

liggitt committed:
Make signer admission plugin check on condition update

liggitt committed:
Record Failed condition in signer controller

liggitt committed:
Clean failed CSRs

liggitt committed:
Print and describe failed CSRs

liggitt committed:
Stop waiting for failed CSRs

munnerz commented on pkg/apis/certificates/types.go:
```diff
@@ -72,6 +75,28 @@ type CertificateSigningRequestSpec struct {
+	// as client certs, but that behavior is not standard kubernetes behavior.
+	LegacyUnknownSignerName = "kubernetes.io/legacy-unknown"
+)
+
```
I'd have thought they only need defining here if they are actually used as part of conversion? But you know better than I do with this stuff 😄

liggitt replied:
yes, validation for v1 prevents creation of CSRs with legacy unknown signer and uses the internal API

munnerz commented on pkg/apis/certificates/v1beta1/defaults.go:
```diff
@@ -18,10 +18,12 @@ package v1beta1
 	"strings"
 
 	certificatesv1beta1 "k8s.io/api/certificates/v1beta1"
+	v1 "k8s.io/api/core/v1"
```
nit: don't need `v1` alias

liggitt replied:
goimports adds it automatically

munnerz commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -17,16 +17,44 @@ limitations under the License.
+	// allow adding Approved/Denied conditions
+	allowSettingApproval bool
+	// allow modifying status.certificate
+	allowSettingCertificate bool
```
Ah, I guess this is so one validator structure can be used for all of 'status', 'approval' and the root resource?

liggitt replied:
right... when we call from ValidateStatusUpdate, we allow setting the certificate. When we call from ValidateApprovalUpdate, we don't.

munnerz commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -140,8 +196,128 @@ func ValidateCertificateSigningRequestSignerName(fldPath *field.Path, signerName
+func allowDuplicateConditionTypes(version schema.GroupVersion, oldCSR *certificates.CertificateSigningRequest) bool {
+	switch {
+	case version == certificatesv1beta1.SchemeGroupVersion:
+		return true // compatibility with v1beta1
```
github doesn't have a wide enough pallet of emojis for me to properly express my reaction here

munnerz commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -140,8 +196,128 @@ func ValidateCertificateSigningRequestSignerName(fldPath *field.Path, signerName
+	}
+}
+func hasDuplicateConditionTypes(csr *certificates.CertificateSigningRequest) bool {
+	seen := map[string]bool{}
```
nit: we could just use a `ConditionType` here and avoid the casting?

munnerz commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -140,8 +196,128 @@ func ValidateCertificateSigningRequestSignerName(fldPath *field.Path, signerName
+func allowEmptyConditionType(version schema.GroupVersion, oldCSR *certificates.CertificateSigningRequest) bool {
+	switch {
+	case version == certificatesv1beta1.SchemeGroupVersion:
+		return true // compatibility with v1beta1
```
🤦

munnerz commented on pkg/apis/certificates/validation/validation_test.go:
```diff
@@ -261,7 +263,7 @@ func TestValidateCertificateSigningRequest(t *testing.T) {
 	for name, test := range tests {
 		t.Run(name, func(t *testing.T) {
-			el := ValidateCertificateSigningRequest(&test.csr)
+			el := ValidateCertificateSigningRequest(&test.csr, test.gv)
```
(as a marker/reminder for me for later): needs more tests 😅

liggitt replied:
so many tests added

munnerz commented on staging/src/k8s.io/api/certificates/v1beta1/types.go:
```diff
@@ -132,11 +133,16 @@ type RequestConditionType string
+	// Status of the condition, one of True, False, Unknown.
+	// Approved, Denied, and Failed conditions may not be "False" or "Unknown".
+	// If unset, should be treated as True.
+	Status v1.ConditionStatus `json:"status" protobuf:"bytes,5,opt,name=status,casttype=k8s.io/api/core/v1.ConditionStatus"`
```
This should have `// +optional`

munnerz commented on staging/src/k8s.io/api/certificates/v1beta1/types.go:
```diff
@@ -132,11 +133,16 @@ type RequestConditionType string
 type CertificateSigningRequestCondition struct {
-	// request approval state, currently Approved or Denied.
+	// type of the condition. Known conditions include "Approved", "Denied", and "Failed".
 	Type RequestConditionType `json:"type" protobuf:"bytes,1,opt,name=type,casttype=RequestConditionType"`
```
:eyes: we don't have `// +optional` here - doesn't this mean this field is already required with newer kubectl versions? (but I guess not with API clients)

liggitt replied:
this hasn't changed, so... yes... anything submitting a condition using openapi validation would need to include the field, but they could set it's value to an empty string

liggitt committed:
Update API compatibility fixtures

            compatibility.go:302: json differs
            compatibility.go:303:   strings.Join({
                  	... // 61 identical lines
                  	"      {",
                  	`        "type": "o,c鮽ort昍řČ扷5Ɨ",`,
                + 	`        "status": "ěĂ凗蓏Ŋ蛊ĉy緅縕",`,
                  	`        "reason": "25",`,
                  	`        "message": "26",`,
                - 	`        "lastUpdateTime": "2901-11-14T22:54:07Z"`,
                + 	`        "lastUpdateTime": "1985-03-23T14:10:57Z",`,
                + 	`        "lastTransitionTime": "2352-05-22T04:29:36Z"`,
                  	"      }",
                  	"    ],",
                - 	`    "certificate": "9Q=="`,
                + 	`    "certificate": "cw=="`,
                  	"  }",
                  	"}",
                  }, "\n")

            compatibility.go:308: yaml differs
            compatibility.go:309:   strings.Join({
                  	... // 42 identical lines
                  	`  username: "20"`,
                  	"status:",
                - 	"  certificate: 9Q==",
                + 	"  certificate: cw==",
                  	"  conditions:",
                - 	`  - lastUpdateTime: "2901-11-14T22:54:07Z"`,
                + 	`  - lastTransitionTime: "2352-05-22T04:29:36Z"`,
                + 	`    lastUpdateTime: "1985-03-23T14:10:57Z"`,
                  	`    message: "26"`,
                  	`    reason: "25"`,
                + 	"    status: ěĂ凗蓏Ŋ蛊ĉy緅縕",
                  	"    type: o,c鮽ort昍řČ扷5Ɨ",
                  	"",
                  }, "\n")

            compatibility.go:314: proto differs
            compatibility.go:316:   strings.Join({
                  	... // 62 identical lines
                  	`      3: "26"`,
                  	"      4 {",
                - 	"        1: 29407013647",
                + 	"        1: 480435057",
                  	"        2: 0",
                  	"      }",
                + 	"      5 {",
                + 	"        1: 12066985776",
                + 	"        2: 0",
                + 	"      }",
                + 	`      6: "\304\233\304\202\345\207\227\350\223\217\305\212\350\233\212\304\211y\347\267\205\347\270\225"`,
                  	"    }",
                - 	`    2: "\365"`,
                + 	`    2: "s"`,
                  	"  }",
                  	"}",
                  	... // 3 identical lines
                  }, "\n")

liggitt commented on pkg/apis/certificates/v1beta1/defaults.go:
```diff
@@ -59,18 +67,34 @@ func DefaultSignerNameFromSpec(obj *certificatesv1beta1.CertificateSigningReques
+func ValidateKubeletServingCSR(req *x509.CertificateRequest, usages []certificatesv1beta1.KeyUsage) error {
 	if !reflect.DeepEqual([]string{"system:nodes"}, req.Subject.Organization) {
-		return false
+		return organizationNotSystemNodesErr
```
I think I'll just stick with the first one we hit for now

liggitt commented:
this is ready for another pass

liggitt committed:
Switch issued check to inspect certificate length

munnerz commented on pkg/controller/certificates/cleaner/cleaner.go:
```diff
@@ -158,6 +159,19 @@ func isDeniedPastDeadline(csr *capi.CertificateSigningRequest) bool {
+// isFailedPastDeadline checks if the certificate has a Failed status and the
+// creation time of the CSR is passed the deadline that pending requests are
+// maintained for.
+func isFailedPastDeadline(csr *capi.CertificateSigningRequest) bool {
```
Despite this being quite simple, might be worth including a unit test here given there's already plumbing for it? https://github.com/kubernetes/kubernetes/blob/dcabb7e6a81e7e60a43a82a380bf01fd0604b8f7/pkg/controller/certificates/cleaner/cleaner_test.go

Unless there's an integration test I've missed 😄

liggitt replied:
added a unit test

munnerz commented on staging/src/k8s.io/client-go/util/certificate/csr/csr.go:
```diff
@@ -112,17 +112,24 @@ func WaitForCertificate(ctx context.Context, client certificatesclient.Certifica
+					klog.V(2).Infof("certificate signing request %s is issued", csr.Name)
+					return true, nil
 				}
+				klog.V(2).Infof("certificate signing request %s is approved, waiting to be issued", csr.Name)
```
😢 there are no tests for this function yet.. https://github.com/kubernetes/kubernetes/blob/a333e591b8e1bca4692eca136389de621979c5ed/staging/src/k8s.io/client-go/util/certificate/csr/csr_test.go

Probably a good thing that can be done in a follow up

liggitt replied:
the kubelet cert bootstrap unit tests exercise it

liggitt commented on pkg/apis/certificates/types.go:
```diff
@@ -91,11 +116,17 @@ type RequestConditionType string
 	Type RequestConditionType
+	// Status of the condition, one of True, False, Unknown.
+	// Approved, Denied, and Failed conditions may not be "False" or "Unknown".
+	// If unset, should be treated as True.
```
The docs are also for clients, which can easily read the docs from the latest version and run against an N-1 server.

liggitt commented on staging/src/k8s.io/api/certificates/v1beta1/types.go:
```diff
@@ -134,11 +135,17 @@ type RequestConditionType string
 	Type RequestConditionType `json:"type" protobuf:"bytes,1,opt,name=type,casttype=RequestConditionType"`
+	// Status of the condition, one of True, False, Unknown.
+	// Approved, Denied, and Failed conditions may not be "False" or "Unknown".
+	// If unset, should be treated as True.
```
the new client / old server was what I had in mind. I'll also include the default.

liggitt commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -60,7 +152,71 @@ func ValidateCertificateSigningRequest(csr *certificates.CertificateSigningReque
+
+	if !opts.allowArbitraryCertificate {
+		if err := validateCertificate(csr.Status.Certificate); err != nil {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("status", "certificate"), "", err.Error()))
```
not sure.... when we redact, we've done the following:

```
allErrs = append(allErrs, field.Invalid(dataPath.Key(core.DockerConfigKey), "<secret contents redacted>", err.Error()))
allErrs = append(allErrs, field.Invalid(dataPath.Key(core.DockerConfigJSONKey), "<secret contents redacted>", err.Error()))
```

for things that are impractical to include the entire serialized struct, we've typically just done "":
```
return append(allErrs, field.Invalid(fldPath, "", "cannot set both resource and service backends"))
allErrs = append(allErrs, field.Invalid(fldPath.Child("openAPIV3Schema"), "", err.Error()))
allErrs = append(allErrs, field.Invalid(fldPath, "", fmt.Sprintf("error building validator: %v", err)))
```

liggitt replied:
set to `"<certificate data>"` to make errors more legible

liggitt commented on staging/src/k8s.io/api/certificates/v1beta1/types.go:
```diff
@@ -148,6 +155,11 @@ type CertificateSigningRequestCondition struct {
 	LastUpdateTime metav1.Time `json:"lastUpdateTime,omitempty" protobuf:"bytes,4,opt,name=lastUpdateTime"`
+	// lastTransitionTime is the time the condition last transitioned from one status to another.
+	// If unset, when a new condition type is added or an existing condition's status is changed,
+	// the server defaults this to the current time.
```
probably not? we don't do it for other conditions... the certificates API was somewhat unique in its use of conditions, and this is matching the existing timestamp defaulting for lastUpdateTime

smarterclayton replied:
Sounds reasonable to me too.

liggitt commented:
Comments addressed, delta visible at https://github.com/kubernetes/kubernetes/compare/de2ad77dc8afd3604de18cd6ae45a6644219fdb8..01a835ae5da08076c0e0cbf2c32c798884cf2d2d

liggitt commented:
/hold for API review from @smarterclayton

munnerz commented:
/lgtm

liggitt commented:
regenerated openapi

liggitt commented:
/retest

smarterclayton commented on pkg/apis/certificates/v1beta1/defaults.go:
```diff
@@ -59,18 +67,34 @@ func DefaultSignerNameFromSpec(obj *certificatesv1beta1.CertificateSigningReques
 	}
 }
 
+var (
```
So much better

smarterclayton commented:
API changes look fine to me, lgtm from that perspective.

liggitt commented:
thanks

/hold cancelk

smarterclayton commented on pkg/controller/certificates/signer/signer.go:
```diff
@@ -153,9 +154,21 @@ func (s *signer) handle(csr *capi.CertificateSigningRequest) error {
+			return fmt.Errorf("error adding failure condition for csr: %v", err)
+		}
+		return nil
+	} else if !recognized {
```
Under the assumption that another controller might be implementing those right now, but we don't need to know about it?

liggitt replied:
right

smarterclayton commented on pkg/controller/certificates/signer/signer.go:
```diff
@@ -153,9 +154,21 @@ func (s *signer) handle(csr *capi.CertificateSigningRequest) error {
+		}
+		return nil
+	} else if !recognized {
+		// Ignore requests for kubernetes.io signerNames we don't recognize
```
Is there a new unit test for this case?

liggitt replied:
there was an existing test covering this case... I added a unit test exercising the "recognized with error" case

smarterclayton commented:
Only one question, I had to read the KEP again but I didn't see anything really crazy.

smarterclayton commented:
/approve
/lgtm

liggitt commented:
/retest

liggitt commented:
/retest

liggitt commented:
/retest
